**Emmanuel Decena - ME EE**
Deep Learning Assignment

Build a classifier on CIFAR10 dataset using
- 3-layer MLP
- 3-layer +1 CNN

Compare performance

**1. Using 3-Layer MLP**

In [1]:
import tensorflow as tf
import numpy as np

from keras.datasets import cifar10
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers.core import Dense, Flatten, Activation
from keras.layers import Conv2D, MaxPooling2D
from keras.callbacks import ModelCheckpoint
from keras.preprocessing.image import ImageDataGenerator
from keras.utils import to_categorical, plot_model

In [2]:
# load dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()
num_labels = len(np.unique(y_train))
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

image_size = x_train.shape[1]
input_size = image_size * image_size * 3

# resize and normalize
x_train = np.reshape(x_train, [-1, input_size]) 
x_train = x_train.astype('float32') / 255
x_test = np.reshape(x_test, [-1, input_size])
x_test = x_test.astype('float32') / 255

In [3]:
# hyper-parameters
batch_size = 256
n_units = 256
epochs = 30

In [4]:
# model is a 3-layer MLP
model = Sequential(name='3-layer-MLP')
model.add(Dense(n_units, input_dim=input_size, activation='relu', name='input_layer'))
model.add(Dense(n_units, activation='relu', name='hidden_layer'))
model.add(Dense(num_labels, activation='softmax', name='output_layer'))
model.summary()

Model: "3-layer-MLP"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (Dense)          (None, 256)               786688    
_________________________________________________________________
hidden_layer (Dense)         (None, 256)               65792     
_________________________________________________________________
output_layer (Dense)         (None, 10)                2570      
Total params: 855,050
Trainable params: 855,050
Non-trainable params: 0
_________________________________________________________________


In [5]:
model.compile(loss="categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])

history = model.fit(
    x_train,
    y_train,
    validation_data=(x_test, y_test),
    epochs=epochs,
    batch_size=batch_size,
)

loss, acc = model.evaluate(x_test, y_test, batch_size=batch_size, verbose=False)

print("\nTest accuracy: %.1f%%" % (100.0 * acc))


Epoch 1/30
196/196 [==============================] - 6s 30ms/step - loss: 2.0616 - accuracy: 0.2639 - val_loss: 1.9765 - val_accuracy: 0.2836
Epoch 2/30
196/196 [==============================] - 9s 43ms/step - loss: 1.8945 - accuracy: 0.3335 - val_loss: 1.8955 - val_accuracy: 0.3204
Epoch 3/30
196/196 [==============================] - 12s 63ms/step - loss: 1.8316 - accuracy: 0.3606 - val_loss: 1.8244 - val_accuracy: 0.3632
Epoch 4/30
196/196 [==============================] - 7s 35ms/step - loss: 1.7908 - accuracy: 0.3740 - val_loss: 1.7985 - val_accuracy: 0.3627
Epoch 5/30
196/196 [==============================] - 5s 26ms/step - loss: 1.7581 - accuracy: 0.3860 - val_loss: 1.8508 - val_accuracy: 0.3419
Epoch 6/30
196/196 [==============================] - 6s 29ms/step - loss: 1.7315 - accuracy: 0.3947 - val_loss: 1.7339 - val_accuracy: 0.3847
Epoch 7/30
196/196 [==============================] - 6s 31ms/step - loss: 1.7064 - accuracy: 0.4041 - val_loss: 1.7074 - val_accuracy: 0.399

**2. Using 3-Layer CNN**

In [6]:
# load dataset
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

In [7]:
# input image dimensions
image_size = x_train.shape[1]
# resize and normalize
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)
x_train = x_train.astype('float32') / 255
x_test = x_test.astype('float32') / 255

# network parameters
input_shape = (image_size, image_size, 3)
batch_size = 256
kernel_size = 3
pool_size = 3
filters = 32

In [8]:
model = Sequential()
model.add(Conv2D(filters=filters,
                 kernel_size=kernel_size, 
                 activation='relu', 
                 kernel_initializer='he_uniform', 
                 padding='same', 
                 input_shape=input_shape,
                 name='input_layer'))

model.add(MaxPooling2D(pool_size,name='hidden_layer'))
model.add(Flatten(name='hidden_layer2'))
model.add(Dense(10, activation='softmax',name='output_layer'))

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_layer (Conv2D)         (None, 32, 32, 32)        896       
_________________________________________________________________
hidden_layer (MaxPooling2D)  (None, 10, 10, 32)        0         
_________________________________________________________________
hidden_layer2 (Flatten)      (None, 3200)              0         
_________________________________________________________________
output_layer (Dense)         (None, 10)                32010     
Total params: 32,906
Trainable params: 32,906
Non-trainable params: 0
_________________________________________________________________


In [9]:
model.compile(loss="categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])
# train the network
model.fit(x_train, y_train, epochs=epochs, batch_size=batch_size)

loss, acc = model.evaluate(x_test, y_test, batch_size=batch_size)
print("\nTest accuracy: %.1f%%" % (100.0 * acc))

Epoch 1/30
196/196 [==============================] - 22s 110ms/step - loss: 2.0965 - accuracy: 0.2381
Epoch 2/30
196/196 [==============================] - 24s 120ms/step - loss: 1.8773 - accuracy: 0.3403
Epoch 3/30
196/196 [==============================] - 24s 122ms/step - loss: 1.7872 - accuracy: 0.3816
Epoch 4/30
196/196 [==============================] - 26s 131ms/step - loss: 1.7237 - accuracy: 0.4033
Epoch 5/30
196/196 [==============================] - 24s 123ms/step - loss: 1.6778 - accuracy: 0.4217
Epoch 6/30
196/196 [==============================] - 27s 135ms/step - loss: 1.6385 - accuracy: 0.4347
Epoch 7/30
196/196 [==============================] - 32s 166ms/step - loss: 1.6036 - accuracy: 0.4460
Epoch 8/30
196/196 [==============================] - 31s 159ms/step - loss: 1.5734 - accuracy: 0.4577
Epoch 9/30
196/196 [==============================] - 23s 117ms/step - loss: 1.5463 - accuracy: 0.4663
Epoch 10/30
196/196 [==============================] - 19s 97ms/step - lo

###### **3. Performance**

The MLP model was outperformed by the CNN model using the same epochs and layers. 
The MLP's accuracy of ~48% was lower compared to the CNN accuracy of ~55%.
The MLP also used more parameters at 855,050 compared to the CNN using only 32,906. Thus, the CNN model was more parameter efficient and has higher accuracy. Using the CNN model is more suitable for this dataset.